api-report/tree.api.md

```diff
@@ -4,9 +4,34 @@
 
 ```ts
 
+import { Serializable } from '@fluidframework/datastore-definitions';
+
+// @public
+export interface ITreeCursor {
+    down(key: TreeKey, index: number): TreeNavigationResult;
```

Comment:
Having only index-based navigation beneath a key does seem to limit our options w.r.t. performance and tolerating invalidation. With a next/prev interface we could implement a TreeCursor that represented a Place and was robust across insertions and removals of siblings, using the same logic we use for rebasing. Calculating a node's index (a log(n) operation with some representations) is unnecessary with next/prev navigation.

Reply:
That's a good point.  I do plan to add the following helpers:
- downFirst(key) - Select first child of a key (sugar for down(key, 0))
- downLast(key) - Select last child of a key (possible optimization of down(key, length(key) - 1)

I'm considering helpers for selecting the next/prev sibling, but I haven't measured the impact of maintaining the extra state on the cursor's stack.  As you point out, the extra state on the stack could be a win if the binary and/or changeset formats make it hard to convert indices to offsets.

FWIW, I was thinking of the cursor as a short-lived object that does not survive across tree mutations.  This interface is for @nedalhy to read the local store (forest) when building the editable tree we expose to the user, and possibly for reading inserted subtrees from changesets.  Its primary purpose is to abstract the binary and/or changeset encoding.